### PR TITLE
Docs and examples

### PR DESCRIPTION
**Pull request by pbaughman**
Improve the documentation and examples to make it a little easier to get started with apex_launchtest

#### Depends on
 - ~~https://github.com/ApexAI/apex_rostest/pull/6~~
 - ~~https://github.com/ApexAI/apex_rostest/pull/8~~
 - ~~https://github.com/ApexAI/apex_rostest/pull/16~~

#### Todo:
- [x] Have a list of asserts similar to https://docs.python.org/2/library/unittest.html#assert-methods.  ~~Maybe blocked on https://github.com/ApexAI/apex_rostest/issues/7~~
- [x] Describe in more detail how arguments are passed into the launch description and into the tests

---
This change is [Reviewable](https://reviewable.io/reviews/apexai/apex_rostest/10)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,64 @@
-# apex_rostest
-Framework for ROS2 Integration Testing
+# apex_launchtest
```
Comment (dejanpan) [inline README.md:1]: Please add `#Compatibility` section:

`apex_rostest` currently works with ros2 latest (master) release.

Comment (pbaughman) [reply README.md:1]: @dejanpan done

**Comment by pbaughman** [comment]
@dejanpan This is done.  Can you review and approve #16 and then this PR?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,124 @@
...
+Designed to work with [ros2 crystal](https://index.ros.org/doc/ros2/Installation/), but requires a slightly newer version of [launch](https://github.com/ros2/launch).  Use the commit [300196b15b0aa99c045460e7d81333df99b8f8da](https://github.com/ros2/launch/tree/300196b15b0aa99c045460e7d81333df99b8f8da) or newer.
+
+## Quick start example
+Start with the apex_launchtest example [good_proc.test.py](apex_launchtest/examples/good_proc.test.py).  Run the example by doing
+>apex_launchtest apex_launchtest/examples/good_proc.test.py
+
+apex_launchtest will launch the nodes found in the `generate_test_description` function, run the tests from the `TestGoodProcess` class, shut down the launched nodes, and then run the tests from the `TestNodeOutput` class.
+
+#### The Launch Description
+```
```
Comment (dejanpan) [inline README.md]: add py for better syntax highlighting?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,124 @@
...
+            cmd=[path_to_process],
+        ),
+
+        # Start tests right away - no need to wait for anything in this example
+        launch.actions.OpaqueFunction(function=lambda context: ready_fn()),
+    ])
+```
+
+The `generate_test_description` function should return a launch.LaunchDescription object that launches the system to be tested.
+It should also call the ready_fn that is passed in to signal when the tests should start.  In the good_proc.test.py example, there
```
Comment (dejanpan) [inline README.md]: ready_fn => `ready_fn`

File: README.md [hunk 4]
```diff
@@ -1,2 +1,124 @@
...
+is no need to delay the start of the tests so the ready_fn is called concurrently when the launching of the process under test
```
Comment (dejanpan) [inline README.md]: ready_fn => `ready_fn`

File: README.md [hunk 5]
```diff
@@ -1,2 +1,124 @@
...
+
+#### Post-Shutdown Tests
+Any classes that inherit from unittest.TestCase that are decorated with the post_shutdown_test descriptor will be run after the launched
+processes have been shut down.  These tests have access to the exit codes and the stdout of all of the launched processes, as well
+as any data created as a side-effect of running the processes
+
+#### Exit Codes and Standard Out
+the apex_launchtest framework automatically adds some member fields to each test case so that the tests can access process output and exit codes
+
+ * self.proc_info - a [ProcInfoHandler object](apex_launchtest/proc_info_handler)
```
Comment (dejanpan) [inline README.md]: this path is actually "apex_launchtest/apex_launchtest/proc_info_handler.py"

File: README.md [hunk 6]
```diff
@@ -1,2 +1,124 @@
...
+ * self.proc_output - an [IoHandler object](apex_launchtest/io_handler.py)
```
Comment (dejanpan) [inline README.md]: this path is "apex_launchtest/apex_launchtest/io_handler.py"

File: README.md [hunk 7]
```diff
@@ -1,2 +1,124 @@
...
+
+These objects provide dictionary like access to information about the running processes.  They also contain methods that the active tests can
+use to wait for a process to exit or to wait for specific output
+
+## Assertions
+The apex_launchtest framework automatically records all stdout from the launched processes as well as the exit codes from any processes
+that are launched.  This information is made available to the tests via the `proc_info` and `proc_output` object.  These objects can be used
+by one of several assert methods to check the output or exit codes of the process
```
Comment (dejanpan) [inline README.md]: :

File: README.md [hunk 8]
```diff
@@ -1,2 +1,124 @@
...
+Example:
+```
+with assertSequentialStdout(self.proc_output, "proc_name") as cm:
+    cm.assertInStdout("Loop 1")
+    cm.assertInStdout("Loop 2")
+    cm.assertInStdout("Loop 3")
+```
+
+#### Waiting for Output or Exit Codes
+The ActiveTests can also call methods that wait for particular output or a particular process to exit or time out.  These asserts are methods on the proc_output and proc_info objects
```
Comment (dejanpan) [inline README.md]: proc_output => `proc_output`

proc_info => `proc_info`

File: README.md [hunk 9]
```diff
@@ -1,2 +1,124 @@
...
+
+`proc_output.assertWaitFor(msg, proc=None, cmd_args=None, *, strict_proc_matching=True, timeout=10)`
+  - The msg, proc, cmd_args, and strict_proc_matching arguments work the same as the other assert methods.  By default, this method waits on output from any process
+  - timeout:  The amount of time to wait before raising an AssertionError
+
+`proc_info.assertWaitForShutdown(proc, cmd_args=None, *, timeout=10)
```
Comment (dejanpan) [inline README.md]: ` missing at the end

File: README.md [hunk 10]
```diff
@@ -1,2 +1,124 @@
 # apex_launchtest
-Framework for ROS2 Integration Testing
+![build status](https://gitlab.com/ApexAI/apex_rostest/badges/master/build.svg) ![coverage](https://gitlab.com/ApexAI/apex_rostest/badges/master/coverage.svg)
+
+This tool is a framework for ROS2 integration testing using the [ros2 style launch description](https://github.com/ros2/launch/blob/master/ros2launch/examples/example.launch.py).
+It works similarly to rostest, but makes it easier to inspect the processes under test.
+
```
Comment (dejanpan) [inline README.md:11]: Could we maybe list couple things that you can test with this framework, e.g.:
1. the process/node under test started and shutdown correctly
1. two or more nodes under test successfully established ros topic connection
1. the process/node under test died unexpectedly
1. the process/node received the data and published the result on stdout or ros topic
1. the process/node received the right arguments
1. ...

**Review by dejanpan — CHANGES_REQUESTED** [review]

**Comment by dejanpan** [comment]
@pbaughman couple of suggestions after the review.

File: README.md (hunk 9 above)
Comment (pbaughman) [reply README.md]: Oops!

**Comment by pbaughman** [comment]
@dejanpan Ok, I think that addresses your comments

File: README.md (hunk 5 above)
Comment (dejanpan) [reply README.md]: .py extension is missing in the file name

File: README.md [hunk 11]
```diff
@@ -1,2 +1,129 @@
...
-Framework for ROS2 Integration Testing
+![build status](https://gitlab.com/ApexAI/apex_rostest/badges/master/build.svg) ![coverage](https://gitlab.com/ApexAI/apex_rostest/badges/master/coverage.svg)
+
+This tool is a framework for ROS2 integration testing using the [ros2 style launch description](https://github.com/ros2/launch/blob/master/ros2launch/examples/example.launch.py).
+It works similarly to rostest, but makes it easier to inspect the processes under test.  For example
+
+  * The exit codes of all processes are available to the tests.  Tests can check that all processes shut down normally, or with specific exit codes.  Tests can fail when a process dies unexpectedly
+  * The stdout and stderr of all processes are available to the tests.
+  * The command-line used to launch the processes are avilalbe to the tests.
+  * Some tests run concurrnetly with the launch and can interact with the running processes.
```
Comment (dejanpan) [inline README.md]: concurrnetly => concurrently

**Comment by pbaughman** [comment]
Gah!  Ok sorry.  I will fix

**Comment by dejanpan** [comment]
@pbaughman I can also fix these 2 things since it is minor?

**Comment by pbaughman** [comment]
@dejanpan nah, I got it

**Review by dejanpan — APPROVED** [review]

**Pull request merged**